File: src/data/4o/Maximum_Sum_of_Subsequence_With_Non-adjacent_Elements/Maximum_Sum_of_Subsequence_With_Non-adjacent_Elements.cpp
#include <vector>
#include <algorithm>
#include <functional>

class Solution {
public:
    int maximumSumSubsequence(vector<int>& nums, vector<vector<int>>& queries) {
        const int MOD = 1e9 + 7;
        int n = nums.size();
        long long totalSum = 0;

        // Function to calculate the maximum sum of a subsequence with non-adjacent elements
        auto maxNonAdjacentSum = [&](const vector<int>& arr) -> long long {
            long long include = 0, exclude = 0;
            for (int num : arr) {
                long long newInclude = exclude + num;
                exclude = max(include, exclude);
                include = newInclude;
            }
            return max(include, exclude);
        };

        // Initial sum for the current nums
        long long currentMaxSum = maxNonAdjacentSum(nums);

        for (auto& query : queries) {
            int pos = query[0];
            int x = query[1];

            // Update the nums array with the new value at pos
            nums[pos] = x;

            // Calculate the new maximum non-adjacent subsequence sum
            currentMaxSum = maxNonAdjacentSum(nums);

            // Add the current maximum to the total sum
            totalSum = (totalSum + currentMaxSum) % MOD;
        }

        return totalSum;
    }
};

//  Test cases for the C++ solution
#include <cassert>

// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Small array with one query
    std::vector<int> nums = {3, 5, 9};
    std::vector<std::vector<int>> queries = {{1, -2}, {0, -3}};
    int expected_result = 21;  // After update nums = [3, 10, 5]; max sum of non-adjacent is 15
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 1 failed.");

    // Test case 2: Multiple queries on the same position
    nums = {0, -1};
    queries = {{0, -5}};
    expected_result = 0;  // No non-adjacent subsequence gives sum > 0
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 2 failed.");

    // Test case 3: Alternating updates on larger array
    nums = {3, 2, 1, 6, 4};
    queries = {{0, 7}, {3, 2}, {4, 9}};
    expected_result = 42;  // After updates, expected sum
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 3 failed.");

    // Test case 4: All elements zero initially, single update
    nums = {0, 0, 0};
    queries = {{2, 5}};
    expected_result = 5;  // After update nums = [0, 0, 5]; max sum is 5
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 4 failed.");

    // Test case 5: Single element update with zero value
    nums = {6, 5, 8};
    queries = {{1, 0}};
    expected_result = 14;  // After update nums = [6, 0, 8]; max sum is 14
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 5 failed.");

    // Test case 6: Larger array with alternating high values
    nums = {5, 10, 5, 10, 5};
    queries = {{1, 5}, {3, 5}, {4, 20}};
    expected_result = 60;  // After updates: expected sum
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 6 failed.");

    // Test case 7: No queries (should handle empty case)
    nums = {4, 6, 8, 10};
    queries = {};
    expected_result = 0;  // No updates; sum of answers = 0
    assert(solution.maximumSumSubsequence(nums, queries) == expected_result && "Test case 7 failed.");
}

int main() {
    Solution solution;
    for (int i = 0; i < 1310720; ++i) {
        testSolution(solution);  
    }
    return 0;
}

File: src/data/o1/Maximum_Sum_of_Subsequence_With_Non-adjacent_Elements/Maximum_Sum_of_Subsequence_With_Non_adjacent_Elements.java
class Solution {
    public int maximumSumSubsequence(int[] nums, int[][] queries) {
        long total = 0;
        final int MOD = 1_000_000_007;
        for(int[] query : queries){
            int pos = query[0];
            int xi = query[1];
            nums[pos] = xi;
            long maxSum = maxSumNoAdjacent(nums);
            total = (total + maxSum) % MOD;
        }
        return (int)total;
    }
    
    private long maxSumNoAdjacent(int[] nums){
        long include = 0;
        long exclude = 0;
        for(int num : nums){
            long newInclude = exclude + num;
            long newExclude = Math.max(include, exclude);
            include = newInclude;
            exclude = newExclude;
        }
        return Math.max(include, exclude);
    }
}

public class Maximum_Sum_of_Subsequence_With_Non_adjacent_Elements {
public static void runTests(Solution solution) {
    // Test case 1: Small array with one query
    int[] nums1 = {3, 5, 9};
    int[][] queries1 = {{1, -2}, {0, -3}};
    int expected_result1 = 21;
    assert solution.maximumSumSubsequence(nums1, queries1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: Multiple queries on the same position
    int[] nums2 = {0, -1};
    int[][] queries2 = {{0, -5}};
    int expected_result2 = 0;
    assert solution.maximumSumSubsequence(nums2, queries2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: Alternating updates on larger array
    int[] nums3 = {3, 2, 1, 6, 4};
    int[][] queries3 = {{0, 7}, {3, 2}, {4, 9}};
    int expected_result3 = 42;
    assert solution.maximumSumSubsequence(nums3, queries3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: All elements zero initially, single update
    int[] nums4 = {0, 0, 0};
    int[][] queries4 = {{2, 5}};
    int expected_result4 = 5;
    assert solution.maximumSumSubsequence(nums4, queries4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Single element update with zero value
    int[] nums5 = {6, 5, 8};
    int[][] queries5 = {{1, 0}};
    int expected_result5 = 14;
    assert solution.maximumSumSubsequence(nums5, queries5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Larger array with alternating high values
    int[] nums6 = {5, 10, 5, 10, 5};
    int[][] queries6 = {{1, 5}, {3, 5}, {4, 20}};
    int expected_result6 = 60;
    assert solution.maximumSumSubsequence(nums6, queries6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: No queries (should handle empty case)
    int[] nums7 = {4, 6, 8, 10};
    int[][] queries7 = {};
    int expected_result7 = 0;
    assert solution.maximumSumSubsequence(nums7, queries7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 1000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    